src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/HashOutputStream.java

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.bazel.repository.downloader;
+
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hasher;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.ThreadCompatible;
+import java.io.IOException;
+import java.io.OutputStream;
+import javax.annotation.Nullable;
+import javax.annotation.WillCloseWhenClosed;
+
+/**
+ * Output stream that guarantees its contents matches a hash code.
+ *
+ * <p>The actual checksum is computed gradually as the output is written. If it
+ * doesn't match, then an {@link IOException} will be thrown when {@link #close()}
+ * is called. This error will be thrown multiple times if these methods are called
+ * again for some reason.
+ *
+ * <p>Note that as the checksum can only be computed once the stream is closed,
+ * data will be written to the underlying stream regardless of whether it matches
+ * the expected checksum.
+ *
+ * <p>This class is not thread safe, but it is safe to message pass this object
+ * between threads.
+ */
+@ThreadCompatible
+public final class HashOutputStream extends OutputStream {
+
+  private final OutputStream delegate;
+  private final Hasher hasher;
+  private final HashCode code;
```

Comment:
It looks like "code" is the passed-in code this checks against, and "actual" is the computed checksum to compare with? Add some comments to clarify this (and to explain why this needs to be "volatile", because I'm not sure why).

Reply:
These are also copied from `HashInputStream.java`. I have no context on why these names were chosen, or what the purpose of `volatile` is for the `Hash*Stream` classes.

To clarify: I'm happy to _change_ these to be something else, but I don't have the ability to explain why existing Bazel code is the way it is.

Reply:
Fair point. Let's leave as is, then.